File: app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/ThumbnailPreviewHandler.java
package org.jellyfin.androidtv.ui.playback.overlay;

import android.graphics.Bitmap;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;

import androidx.leanback.widget.PlaybackSeekDataProvider;

import org.jellyfin.androidtv.R;

public class ThumbnailPreviewHandler {

    private final PlaybackSeekDataProvider seekProvider;
    private final ImageView thumbnailView;
    private final long skipForwardLength;

    public ThumbnailPreviewHandler(PlaybackSeekDataProvider seekProvider,
                                   ImageView thumbnailView,
                                   long skipForwardLength) {
        this.seekProvider = seekProvider;
        this.thumbnailView = thumbnailView;
        this.skipForwardLength = skipForwardLength;
    }

    public void updateThumbnailPreview(long previewSeekPosition) {
        if (seekProvider == null || thumbnailView == null) return;

        int thumbnailIndex = (int) (previewSeekPosition / skipForwardLength);

        PlaybackSeekDataProvider.ResultCallback thumbnailCallback = new PlaybackSeekDataProvider.ResultCallback() {
            @Override
            public void onThumbnailLoaded(Bitmap bitmap, int index) {
                if (bitmap == null) return;

                thumbnailView.setImageBitmap(bitmap);
                showThumbnailPreview();
            }
        };

        seekProvider.getThumbnail(thumbnailIndex, thumbnailCallback);
    }

    public void showThumbnailPreview() {
        if (thumbnailView == null) return;

        thumbnailView.setVisibility(View.VISIBLE);
    }

    public void hideThumbnailPreview() {
        if (thumbnailView == null || thumbnailView.getVisibility() != View.VISIBLE) return;

        Animation fadeOut = AnimationUtils.loadAnimation(thumbnailView.getContext(), R.anim.fade_out);
        fadeOut.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {}

            @Override
            public void onAnimationEnd(Animation animation) {
                thumbnailView.setVisibility(View.GONE);
                thumbnailView.setImageBitmap(null); // Clear the bitmap to free memory
            }

            @Override
            public void onAnimationRepeat(Animation animation) {}
        });

        thumbnailView.startAnimation(fadeOut);
    }

    public boolean isAvailable() {
        return seekProvider != null && thumbnailView != null;
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/ThumbnailPreviewHandler.kt
package org.jellyfin.androidtv.ui.playback.overlay

import android.graphics.Bitmap
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.leanback.widget.PlaybackSeekDataProvider
import org.jellyfin.androidtv.ui.composable.ThumbnailPreview
import java.util.concurrent.atomic.AtomicInteger

class ThumbnailPreviewHandler(
    private val seekProvider: PlaybackSeekDataProvider?,
    private val thumbnailComposeView: ComposeView?
) {
    private var currentBitmap by mutableStateOf<Bitmap?>(null)
    private val requestNumber = AtomicInteger(0)
    private val lastShownRequestNumber = AtomicInteger(0)

    companion object {
        private const val MAX_REQUEST_AGE = 20
    }

    init {
        thumbnailComposeView?.setContent {
            ThumbnailPreview(bitmap = currentBitmap)
        }
    }

    fun updateThumbnailPreview(previewSeekPosition: Long) {
        val currentRequestNumber = requestNumber.incrementAndGet()
        
        val thumbnailCallback = object : PlaybackSeekDataProvider.ResultCallback() {
            override fun onThumbnailLoaded(bitmap: Bitmap?, index: Int) {
                if (bitmap == null) return
                
                // Ignore requests that are older than what we last showed
                if (currentRequestNumber <= lastShownRequestNumber.get()) return
                
                // Ignore requests that are too old
                if (currentRequestNumber < requestNumber.get() - MAX_REQUEST_AGE) return
                
                lastShownRequestNumber.set(currentRequestNumber)
                showThumbnail(bitmap)
            }
        }

        val thumbnailIndex = findClosestIndex(seekProvider?.seekPositions ?: LongArray(0), previewSeekPosition)
        seekProvider?.getThumbnail(thumbnailIndex, thumbnailCallback)
    }

    private fun findClosestIndex(positions: LongArray, targetPosition: Long): Int {
        if (positions.isEmpty() || targetPosition < 0) return 0
        
        return positions.indexOfFirst { it >= targetPosition }
            .takeIf { it >= 0 } ?: (positions.size - 1).coerceAtLeast(0)
    }

    fun showThumbnail(bitmap: Bitmap) {
        currentBitmap = bitmap
    }

    fun hideThumbnailPreview() {
        currentBitmap = null 
    }

    fun isAvailable(): Boolean {
        return seekProvider != null && thumbnailComposeView != null
    }
} 